### ISSUE TITLE
Refresh GraphQL data after creating pages programmatically

### ISSUE THREAD

**Issue by beliayeu** [gatsbyjs/gatsby#19934]
## Description

given there is a template that contains a graphql query
given there are no pages created with this template
given gatsby-source-filesystem plugin is configured
given new pages are created on adding new files to local file system
given gatsby develop is running
when a new file is added to filesystem and a page created for this file programmatically
when the page is opened in the browser
then there is a TypeError

### Steps to reproduce

1. take https://github.com/beliayeu/gatsby-starter-default/tree/bug/refresh-graphql-on-page-create
2. yarn
3. yarn develop
4. touch src/data/test.txt
5. open http://localhost:8000/test

### Expected result

no errors. page, created from the template, is rendered

### Actual result

TypeError: Cannot read property  of undefined.

Note: the problem is gone, after restarting gatsby server with clearing cache

### Environment

  System:
    OS: macOS 10.15
    CPU: (4) x64 Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
    Shell: 3.2.57 - /bin/bash
  Binaries:
    Node: 10.15.3 - /usr/local/bin/node
    Yarn: 1.15.2 - /usr/local/bin/yarn
    npm: 6.13.1 - /usr/local/bin/npm
  Languages:
    Python: 2.7.10 - /usr/local/bin/python
  Browsers:
    Chrome: 78.0.3904.108
    Safari: 13.0.2
  npmPackages:
    gatsby: 2.18.6 => 2.18.6
    gatsby-source-filesystem: ^2.1.40 => 2.1.40
  npmGlobalPackages:
    gatsby-cli: 2.7.28

**Comment by beliayeu** [comment]
the workaround I tried is to execute updateStateAndRunQueries on createPages phase. see [changes](https://github.com/beliayeu/gatsby-starter-default/compare/bug/refresh-graphql-on-page-create...beliayeu:bugfix/update-state-and-run-queries-on-page-create?expand=1)

**Comment by vladar** [comment]
**tldr**; the simple workaround, for now, is to have at least one page created for a template before starting `gatsby develop`.

Why this bug occurs (the extended version):

We extract page queries only for components, previously registered in redux `state.components`:

https://github.com/gatsbyjs/gatsby/blob/da9e273560ca352747045a540564c4ad9427f3ed/packages/gatsby/src/query/query-watcher.js#L111-L117

And even though a query for our template is parsed and exists in `queries` map above, it won't be registered with `boundActionCreators.queryExtracted` because there is no component for it.

And there is no component because we add components to state on `CREATE_PAGE` action. There are no pages hence the component is not in the state.

So what happens when we create the first page? The page component is being added to state but with `component.query === ""`. As a result, the query for this page is not executed.

Not sure what should be the right fix for this. As for me, we should store components with parsed queries separately, independent of pages. But there might be many corner cases here which I am not aware of. We will discuss it with @pieh after the holidays.

**Comment by LekoArts** [comment]
Closing this as stale since in the meantime Gatsby v3 and updated related packages were released. Please try with the latest versions and if you still see this problem open a new bug report (it must include a [minimal reproduction][reproduction]).

[reproduction]: https://gatsby.dev/reproduction
